resources/js/components/shared/TimelinePlot.vue

```diff
@@ -0,0 +1,393 @@
+      d3.select(".resize e").select("rect").attr("height",height);
+      d3.select(".background").attr("height",height);
+
+      // general function to update the chart based on specified boundaries
+      function updateChart(new_x1, new_x2, delta_t) {
+        return function() {
+          // update x-axis
+          x.domain([new_x1, new_x2]);
+          xAxis.transition().duration(delta_t)
+            .call(d3.svg.axis().scale(x).orient("bottom"));
+          for (let i = 0; i < data.length; i += 1) {
+            // update line position
+            plot_lines[i].transition().duration(delta_t)
+              .attr("d", d3.svg.line()
+                .x(getXRange)
+                .y(getYRange)
+              );
+            // update circle positions
+            plot_points[i].transition().duration(delta_t)
+              .attr("cx", getXRange)
+              .attr("cy", getYRange);
+          }
+        };
+      }
+
+      // zoom-in the plot using the coordinates of the brushed area
+      function zoomOnSelection() {
+        // get the boundaries of the selection
+        let extent = d3.select(".extent");
+        let coord1 = Number(extent.attr("x"));
+        let coord2 = coord1 + Number(extent.attr("width"));
+
+        if (coord2 - coord1 < 2) {
+          return; // selection too small to be meaningfull (e.g. accidental mouse click)
+        } else {
+          updateChart(x.invert(coord1), x.invert(coord2), 700)();
+        }
+      }
+
+      // on double-click, reinitialize the chart to original zoom level
+      d3.select("#timeline_plot").on("dblclick", updateChart(x_min, x_max, 200))
+
+
+      // ################## ADD TOOLTIP ##################
+      // add div which serves as the tooltip
+      let tooltip = d3.select("#timeline_plot").append("div")
+        .attr("class", "my_tooltip")
+        .style("position", "absolute")
+        .style("text-align", "left")
+        .style("opacity", "0")
+        .style("background-color", "white")
+        .style("border", "solid")
+        .style("border-width", "1px")
+        .style("border-radius", "3px")
+        .style("padding", "5px");
+
+      // make tooltip visible when user hovers on a point
+      function showTooltip(d,i) {
+        d3.select(this).style("opacity", "1"); // make circle appear
+        tooltip.transition().duration(100).style("opacity", "1");
+      }
+
+      // load information about data point and display it in the tooltip div
+      function renderTooltip(k) {
+        return function(d,i) {
+          // get x,y coordinates from the circle the user is hovering on
+          let [x_val, y_val] = Object.values(d).map(String);
+          // render the html div element with this point's data
+          let content =
+            `<b>${line_names[k]}</b>: ${y_val}
+            <br/>
+            <b>${labels.x_axis}</b>: ${x_val}`;
+          tooltip.html(content) // add slight offset for visiblity
```

Comment:
That's a great way to introduce an XSS vulnerability.  By allowing (potentially) user-controlled content to get inserted into HTML, a malicious user could potentially inject a script tag and execute malicious JS.  If you can use Vue to insert the text, that would be better.  If you can't do that easily, it would be better to use jQuery's `.text()` function.